moveit_ros/moveit_servo/src/servo_calcs.cpp

```diff
@@ -732,60 +770,51 @@ double ServoCalcs::velocityScalingFactorForSingularity(const Eigen::VectorXd& co
   return velocity_scale;
 }
 
-void ServoCalcs::enforceVelLimits(Eigen::ArrayXd& delta_theta)
+bool ServoCalcs::enforceVelAccelLimitsWithReflexxes(Eigen::ArrayXd& delta_theta)
 {
-  Eigen::ArrayXd velocity = delta_theta / parameters_->publish_period;
-
-  std::size_t joint_delta_index = 0;
-  // Track the smallest velocity scaling factor required, across all joints
-  double velocity_limit_scaling_factor = 1;
-
-  for (auto joint : joint_model_group_->getActiveJointModels())
+  // Update current state in Reflexxes
+  // For current state, we know the current position well. For vel and accel, use the previous output from Reflexxes
+  // This assumes the hardware tracks the requested vel/accel well.
```

Comment:
Yes, a rather strong assumption.

I've run into this before and it doesn't always work.

Reply:
OK, I understand. Not sure I will change it right now.

BTW I saw the 2 posts on ROS Discourse about jerk-limited smoothing. I would love to use one of those but unfortunately this PR was 90% done when I saw the posts. At least this PR will serve as a template, if somebody (maybe me!) wants to use one of those libraries instead.

Another thing I saw is that Reflexxes tends to overshoot depending on the acceleration limit, its current velocity, and target velocity. Basically it needs to overshoot to give itself more time to accelerate to the target velocity. This is pretty counterintuitive and might be a surprise for many people. The robot can even more backward if the wrong control frequency is used.

It takes careful parameter tuning. I found that running at a lower control frequency helped.

Reply:
I'm interested to know which of these _you_ think moveit_servo should use:

- Reflexxes II
- Simple lowpass filter
- opt_control
- pantor/ruckig

The lowpass filter has a lot of benefits. It's basically idiot-proof but it doesn't work on some robots.

Reply:
@AndyZe: you could use this assumption only for robots not providing velocity and acceleration in their `joint_state` message. Or what you think about calculating those from positions? (I know that derivative are not so simple to calculate, but with sufficient frequency, they should be sufficiently correct).

Reply:
> I'm interested to know which of these *you* think moveit_servo should use:

I don't really have a preference. Getting `moveit_servo` to output jerk limited control signals would be great, so anything which does that would seem to be acceptable.

My comment was really just about the comment I commented on (hah).

It's a rather strong assumption which is easily violated, and it's buried somewhere in the code.

@destogl wrote:
> Or what you think about calculating those from positions? (I know that derivative are not so simple to calculate, but with sufficient frequency, they should be sufficiently correct).

I've seen some really weird things happen with derivatives and reflexxes / other trajectory generators.